services/api/service.go

```diff
@@ -484,94 +486,144 @@ func (api *RelayAPI) handleRegisterValidator(w http.ResponseWriter, req *http.Re
 		"mevBoostV": common.GetMevBoostVersionFromUserAgent(ua),
 	})
 
-	respondError := func(code int, msg string) {
-		log.Warn("bad request: ", msg)
-		api.RespondError(w, code, msg)
-	}
-
 	start := time.Now()
 	registrationTimeUpperBound := start.Add(10 * time.Second)
 
-	registrations := []types.SignedValidatorRegistration{}
+	numRegTotal := 0
+	numRegProcessed := 0
+	numRegActive := 0
 	numRegNew := 0
+	processingStoppedByError := false
 
-	if err := json.NewDecoder(req.Body).Decode(&registrations); err != nil {
-		respondError(http.StatusBadRequest, "failed to decode payload")
+	respondError := func(code int, msg string) {
+		processingStoppedByError = true
+		log.Warnf("error: %s", msg)
+		api.RespondError(w, code, msg)
+	}
+
+	if req.ContentLength == 0 {
+		respondError(http.StatusBadRequest, "empty request")
+		return
+	}
+
+	body, err := io.ReadAll(req.Body)
+	if err != nil {
+		log.WithField("contentLength", req.ContentLength).Warn("failed to read request body")
+		api.RespondError(w, http.StatusBadRequest, "failed to read request body")
 		return
 	}
 
-	// Possible optimisations:
-	// - GetValidatorRegistrationTimestamp could keep a cache in memory for some time and check memory first before going to Redis
-	// - Do multiple loops and filter down set of registrations, and batch checks for all registrations instead of locking for each individually:
-	//   (1) sanity checks, (2) IsKnownValidator, (3) CheckTimestamp, (4) Batch SetValidatorRegistration
-	for _, registration := range registrations {
-		if registration.Message == nil {
-			respondError(http.StatusBadRequest, "registration without message")
+	parseRegistration := func(value []byte) (pkHex types.PubkeyHex, timestampInt int64, err error) {
+		pubkey, err := jsonparser.GetUnsafeString(value, "message", "pubkey")
+		if err != nil {
+			return pkHex, timestampInt, fmt.Errorf("registration message error (pubkey): %w", err)
+		}
+
+		timestamp, err := jsonparser.GetUnsafeString(value, "message", "timestamp")
+		if err != nil {
+			return pkHex, timestampInt, fmt.Errorf("registration message error (timestamp): %w", err)
+		}
+
+		timestampInt, err = strconv.ParseInt(timestamp, 10, 64)
+		if err != nil {
+			return pkHex, timestampInt, fmt.Errorf("invalid timestamp: %w", err)
+		}
+
+		return types.PubkeyHex(pubkey), timestampInt, nil
+	}
+
+	// Iterate over the registrations
+	_, err = jsonparser.ArrayEach(body, func(value []byte, dataType jsonparser.ValueType, offset int, _err error) {
+		numRegTotal += 1
+		if processingStoppedByError {
 			return
 		}
```

Comment:
I think it makes sense to stop processing if one of the registrations has a problem, but is there a situation where this isn't the desired behavior? Previously, it would skip the bad registration and continue processing.

Reply:
Previously it would also abort on a bad registration, see here: https://github.com/flashbots/mev-boost-relay/blob/main/services/api/service.go#L508-L530

Reply:
This is the desired behavior, because otherwise an attacker could send 10k invalid validator registrations, each taking 1ms+ for signature validation, thus totally exhausting the resources.

Reply:
Oops, you're right, it would abort previously. Yes, that makes sense!

Reply:
@metachris regarding the possible attack you mentioned, is there anything preventing duplicate registrations in the list of registrations? If not, you could send a list of 10k duplicate & valid registrations. They could have the same timestamp, but it would be more impactful if each incremented the timestamp by one, like an index. That way it would cause a database write. This would effectively bypass the relay's IP rate limiting. Thoughts?

Reply:
Any known registration is checked and skipped before signature validation already, so even 10k known registrations wouldn't add a significant overhead, i think. perhaps there's further optimisations 🤔 

May be good to open a Github issue about possible optimisations here.